llmfoundry/models/hf/hf_causal_lm.py

```diff
@@ -77,196 +76,209 @@ def __init__(self, om_model_config: Union[DictConfig,
+
+        peft_config = pop_config(om_model_config,
+                                 'peft_config',
+                                 must_exist=False,
+                                 convert=True)
+
+        if peft_config is not None:
+            if peft_installed:
+                from peft import LoraConfig
+                peft_type = peft_config.get('peft_type', None)
+                if peft_type.upper() != 'LORA':
                     raise ValueError(
-                        f'attention_patch_type is only supported for llama models, but got {model.config.model_type}'
+                        f'Only LORA is supported for peft_type, but got {peft_type}.'
                     )
-
-                log.debug(
-                    f'Patching llama attention with {attention_patch_type} attention'
+                task_type = peft_config.get('task_type', None)
+                if task_type.upper() != 'CAUSAL_LM':
+                    raise ValueError(
+                        f'Only CAUSAL_LM is supported for task_type, but got {task_type}.'
+                    )
+                peft_config = LoraConfig(**peft_config)
+            else:
+                raise ValueError(
+                    'PEFT is not installed, but peft_config was passed. Please install LLM Foundry with the peft extra to use peft_config.'
                 )
-                from transformers.models.llama.modeling_llama import \
-                    LlamaAttention
-
-                from llmfoundry.models.layers.llama_attention_monkeypatch import \
-                    get_llama_attention_patch_fn
-                LlamaAttention.forward = get_llama_attention_patch_fn(
-                    attention_patch_type)
-                model.config.use_cache = False
-
-        # elif the model is either a PeftModel or a PreTrainedModel
-        elif isinstance(om_model_config, model_types):
-            model = om_model_config
-            init_device = 'cpu'
-            z_loss = 0.0
-
-        # else, unsupported type
-        else:
-            raise ValueError(
-                f'om_model_config must be either a DictConfig, PeftModel, or PreTrainedModel, but got {type(om_model_config)}'
-            )
 
-        composer_model = super().__init__(model=model,
-                                          shift_labels=True,
-                                          tokenizer=tokenizer,
-                                          metrics=train_metrics,
-                                          eval_metrics=eval_metrics,
-                                          z_loss=z_loss,
-                                          init_device=init_device)
+        composer_model = super().__init__(
+            model=model,
+            shift_labels=True,
+            tokenizer=tokenizer,
+            metrics=train_metrics,
+            eval_metrics=eval_metrics,
+            z_loss=z_loss,
+            init_device=init_device,
+            peft_config=peft_config,
```

Comment:
is it possible to refactor such that where appropriate there's one helper per named argument?

Reply:
plus also a separate helper for patching attention would be nice

Reply:
did a bit of refactoring. I'm sure there is more that could be done, but I'm gonna say that is out of scope for this PR :) Please review again though as I moved a bunch of stuff around.